### ISSUE TITLE
Ship with an internal proxy for fully offline builds?

### ISSUE THREAD

**Issue by amtlib-dot-dll** [flatpak/flatpak-builder#110]
During my trial to build VS Code offline, I discovered many dependent packages have mandatory need to accessing the Internet like [`vscode-ripgrep`](https://github.com/roblourens/vscode-ripgrep/blob/master/lib/postinstall.js) and [`vscode-extension-vscode`](https://github.com/Microsoft/vscode-extension-vscode/blob/master/bin/install). They download files during the post-installation steps. It is too annoying to patch up these packages one by one.

If the build environment have very limited access to the Internet that is specified in a list of hash-verified files, it would be much better.

**Comment by TingPing** [comment]
While that is certainly one solution, it being a part of `flatpak-builder` to me gives the impression that these awful build systems are acceptable. Every project should be able to build offline. If that requires automated tooling that is still reasonable but it should never just download everything.

See also: https://github.com/flatpak/flatpak-builder-tools

**Comment by amtlib-dot-dll** [comment]
Yes, that _is_ awful. But complaining won't help. IMHO in the JavaScript ecosystem, rapid iteration is more important than stability and reproducibility, because even the best programs will be replaced and discarded in a few moments. `flatpak-builder` currently supports only C-relative build systems :wink: but there are many applications containing no C code lines at all

**Comment by amtlib-dot-dll** [comment]
The build system of VS Code is `yarn` and it does support offline building in first class. However some dependent packages connect to Internet and complain when disconnected. You are right, `npm` is a bad build system, because those packages connect the internet just to some conditional downloads that `npm` cannot do. And this also burdens `yarn`.

**Comment by bbhtt** [comment]
Seems out of scope to me.
